### PR TITLE
ci-operator multi-stage: gather artifacts

### PR DESCRIPTION
**Pull request by bbguimaraes**
When given an output artifact directory, for each step that declares an
artifact directory, copy files from the container to
`$artifact_dir/$test_name/$step_name/`.

---

Tests involving `pkg/steps/artifacts.go` are impractical and it would be a lot
of work to invent a separate test infrastructure, so tests are limited to pod
generation. I am open for ideas, though, and to working on that in a separate
PR.

### CODE REVIEW

File: pkg/steps/multi_stage.go [hunk 1]
```diff
@@ -145,6 +154,13 @@ func (s *multiStageTestStep) generatePods(steps []api.TestStep) ([]coreapi.Pod,
 		if owner := s.jobSpec.Owner(); owner != nil {
 			pod.OwnerReferences = append(pod.OwnerReferences, *owner)
 		}
+		if s.artifactDir != "" && step.ArtifactDir != "" {
```
Comment (stevekuznetsov) [inline pkg/steps/multi_stage.go]: Why do we need to check both?

File: pkg/steps/multi_stage_test.go [hunk 2]
```diff
@@ -118,7 +124,7 @@ func TestGeneratePods(t *testing.T) {
 		},
 		Namespace: "namespace",
 	}
-	step := newMultiStageTestStep(config.Tests[0], &config, nil, &jobSpec)
+	step := newMultiStageTestStep(config.Tests[0], &config, nil, "artifact_dir", &jobSpec)
```
Comment (stevekuznetsov) [inline pkg/steps/multi_stage_test.go]: Where is this value used?

File: pkg/steps/multi_stage.go (hunk 1 above)
Comment (bbguimaraes) [reply pkg/steps/multi_stage.go]: `s.artifactDir`: the user has requested artifact gathering via `--artifact-dir`, which is where this value ultimately comes from. There is nowhere to output to if that's not the case.

`step.ArtifactDir`: the step has declared an artifact directory. There is no reason to add the mount if it doesn't.

File: pkg/steps/multi_stage_test.go (hunk 2 above)
Comment (bbguimaraes) [reply pkg/steps/multi_stage_test.go]: Nowhere, but it triggers the conditional in your previous comment.

**Comment by bbguimaraes** [comment]
I'm not sure what you mean by "concatenation behavior", but I've explained why there are not tests for the actual artifact gathering in the description.

File: pkg/steps/multi_stage.go (hunk 1 above)
Comment (bbguimaraes) [reply pkg/steps/multi_stage.go]: The pod step (unbeknownst to me when I wrote the multi-stage code) uses the exact same logic: https://github.com/openshift/ci-tools/blob/36ac9f630ea7308baacfba2779564d6f5cf544eb/pkg/steps/pod.go#L131.

Comment (bbguimaraes) [reply pkg/steps/multi_stage.go]: Although maybe we should add the volume even if `--artifact-dir` is not specified? If a test has `artifact_dir`, I would think it requires that directory to exist.

But… it has worked so far, so /shrug?

**Pull request merged**